core/model/impl/src/main/kotlin/parsing/BooleanExpressionParser.kt

```diff
@@ -0,0 +1,302 @@
+                    }
+                }
+
+                UnaryNot, BinaryAnd, BinaryOr -> {
+                    while (stack.lastOrNull().let { top ->
+                            when (val kind = top?.kind) {
+                                // Operators
+                                BinaryAnd, BinaryOr -> token.kind.ordinal >= kind.ordinal
+                                UnaryNot -> token.kind != UnaryNot
+                                else -> false
+                            }
+                        }) {
+                        inversePostfix += stack.removeLast()
+                    }
+                    stack += token
+                }
+
+                Variable -> {
+                    inversePostfix += token
+                }
+
+                EndOfExpression -> while (stack.isNotEmpty()) {
+                    val top = stack.removeLast()
+                    if (top.kind == LeftParenthesis) {
+                        throw ParseException(
+                            source = expressionSource,
+                            where = top.sourceLocation,
+                            what = "Unclosed $top",
+                        )
+                    }
+                    inversePostfix += top
+                }
+            }
+        }
+
+        return inversePostfix
+    }
+
+    private fun buildExpressionTree(inversePostfix: List<Token>): E {
+        val expressionStack = arrayListOf<E>()
+        for (token in inversePostfix) {
+            when (token.kind) {
+                UnaryNot -> expressionStack.add(factory.createNot(expressionStack.removeLast()))
+                BinaryAnd -> expressionStack.add(factory.createAnd(
+                    rhs = expressionStack.removeLast(),
+                    lhs = expressionStack.removeLast(),
+                ))
+
+                BinaryOr -> expressionStack.add(factory.createOr(
+                    rhs = expressionStack.removeLast(),
+                    lhs = expressionStack.removeLast(),
+                ))
+
+                Variable -> {
+                    when (val result = factory.parseVariable(token.text)) {
+                        is Factory.ParseResult.Error -> throw ParseException(
+                            source = expressionSource,
+                            where = token.sourceLocation..<token.sourceLocation + token.text.length - 1,
+                            what = "Variable: ${result.message}"
+                        )
+
+                        is Factory.ParseResult.Ok ->
+                            expressionStack.add(result.variable)
+                    }
+                }
+
+                else -> throw AssertionError("Unexpected token in postfix notation - $token")
+            }
+        }
+        return expressionStack.single()
+    }
+
+    private data class Token(
+        val kind: Kind,
+        val sourceLocation: Int,
+        val text: String,
+    ) {
+        enum class Kind(val description: String) {
```

Comment:
maybe add comment that order affects priority

Reply:
Done